File: lab-J/app/controller/car_controller.py
from flask import Blueprint, abort, redirect, render_template, request, url_for

from app.model.car import Car, find, find_all

bp = Blueprint("car", __name__)


@bp.route("/cars")
def index():
    cars = find_all()
    return render_template("car/index.html", cars=cars)


@bp.route("/cars/<int:id>")
def show(id: int):
    car = find(id)
    if not car:
        abort(404)
    return render_template("car/show.html", car=car)


@bp.route("/cars/create", methods=["GET", "POST"])
def create():
    if request.method == "POST":
        manufacturer = request.form.get("car[manufacturer]")
        model = request.form.get("car[model]")
        color = request.form.get("car[color]")
        year = request.form.get("car[year]")

        car = Car()
        car.manufacturer = manufacturer
        car.model = model
        car.color = color
        try:
            car.year = int(year) if year else None
        except ValueError:
            car.year = None

        car.save()
        return redirect(url_for("car.index"))

    return render_template("car/create.html", car=None)


@bp.route("/cars/<int:id>/edit", methods=["GET", "POST"])
def edit(id: int):
    car = find(id)
    if not car:
        abort(404)

    if request.method == "POST":
        car.manufacturer = request.form.get("car[manufacturer]")
        car.model = request.form.get("car[model]")
        car.color = request.form.get("car[color]")
        year = request.form.get("car[year]")
        try:
            car.year = int(year) if year else None
        except ValueError:
            car.year = None

        car.save()
        return redirect(url_for("car.show", id=car.id))

    return render_template("car/edit.html", car=car)


@bp.route("/cars/delete", methods=["POST"])
def delete():
    id_value = request.form.get("id") or request.form.get("car[id]")
    if not id_value:
        abort(400)

    try:
        id_int = int(id_value)
    except ValueError:
        abort(400)

    car = find(id_int)
    if car:
        car.delete()

    return redirect(url_for("car.index"))

File: lab-I/crud/controllers/carController.js
const Car = require('../models/Car');

function asyncHandler(handler) {
  return function handleAsync(req, res, next) {
    Promise.resolve(handler(req, res, next)).catch(next);
  };
}

async function find(req, res) {
  const car = await Car.find(req.params.id);
  if (!car) {
    res.status(404).render('error', { message: 'Missing car', error: {} });
    return null;
  }
  return car;
}

function parseCarBody(body = {}) {
  return {
    manufacturer: body.manufacturer ? body.manufacturer.trim() : null,
    model: body.model ? body.model.trim() : null,
    color: body.color ? body.color.trim() : null,
    year: body.year ? Number(body.year) : null,
  };
}

async function index(req, res) {
  const cars = await Car.findAll();
  res.render('cars/index', { cars });
}

function createForm(req, res) {
  res.render('cars/create', { car: new Car() });
}

async function create(req, res) {
  const car = Car.fromObject(parseCarBody(req.body));
  await car.save();
  res.redirect('/cars');
}

async function editForm(req, res) {
  const car = await find(req, res);
  if (car) {
    res.render('cars/edit', { car });
  }
}

async function update(req, res) {
  const car = await find(req, res);
  if (car) {
    car.fill(parseCarBody(req.body));
    await car.save();
    res.redirect('/cars');
  }
}

async function show(req, res) {
  const car = await find(req, res);
  if (car) {
    res.render('cars/show', { car });
  }
}

async function remove(req, res) {
  const car = await find(req, res);
  if (car) {
    await car.delete();
    res.redirect('/cars');
  }
}

module.exports = {
  index: asyncHandler(index),
  createForm,
  create: asyncHandler(create),
  editForm: asyncHandler(editForm),
  update: asyncHandler(update),
  show: asyncHandler(show),
  remove: asyncHandler(remove),
};
